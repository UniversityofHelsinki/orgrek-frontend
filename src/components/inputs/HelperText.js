import React from 'react';
import PropTypes from 'prop-types';

const mapLines = (lines) =>
  lines.map((line, index) => (
    <React.Fragment key={`${index}-${line}`}>
      {line}
      {index < lines.length - 1 && <br />}
    </React.Fragment>
  ));

/**
 * Helper text component to be used with MUI TextField helperText prop.
 *
 * Displays errors or a normal helper text. Errors take precedence, so
 * the normal helper text is hidden when errors are present.
 *
 * Both texts can be either one string or an array of strings.
 * When the array contains multiple texts, each text is displayed on separate
 * line.
 */
const HelperText = ({ helperText, errors }) => {
  if (Array.isArray(errors) && errors.length > 0) {
    return mapLines(errors);
  } else if (typeof errors === 'string' && errors !== '') {
    return errors;
  } else if (Array.isArray(helperText) && helperText.length > 0) {
    return mapLines(helperText);
  } else if (typeof helperText === 'string' && helperText !== '') {
    return helperText;
  } else {
    // The space makes sure that the height is always at least one line so that
    // all fields have consistent height whether there is a helper text or not
    return <span aria-hidden="true">&#8203;</span>;
  }
};

HelperText.propTypes = {
  /** Displayed when there are no errors */
  helperText: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),

  /** Displayed in place of helper text if present */
  errors: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),
};

export default HelperText;
