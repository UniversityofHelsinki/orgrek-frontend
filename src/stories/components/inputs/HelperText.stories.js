import HelperText from '../../../components/inputs/HelperText';

export default {
  component: HelperText,
};

export const Default = {
  args: {
    helperText: 'Assistive text',
    errors: [],
  },
};

export const Error = {
  args: {
    helperText: 'Assistive text',
    errors: ['Error text'],
  },
};

export const Multiline = {
  args: {
    helperText: 'Assistive text',
    errors: ['Error 1', 'Error 2'],
  },
};

export const Empty = {
  args: {
    helperText: '',
    errors: [],
  },
};
