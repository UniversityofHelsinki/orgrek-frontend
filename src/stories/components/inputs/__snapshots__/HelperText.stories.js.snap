// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stories/components/inputs/HelperText Default smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Assistive text",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
  ],
  "name": "stories-components-inputs-helpertext--default",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`stories/components/inputs/HelperText Empty smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": undefined,
  "name": "stories-components-inputs-helpertext--empty",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`stories/components/inputs/HelperText Empty smoke-test 2`] = `
Object {
  "checked": undefined,
  "children": undefined,
  "name": "stories-components-inputs-helpertext--empty",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`stories/components/inputs/HelperText Error smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Error text",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
  ],
  "name": "stories-components-inputs-helpertext--error",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`stories/components/inputs/HelperText Multiline smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Error 1",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "
",
      "pressed": undefined,
      "role": "LineBreak",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Error 2",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
  ],
  "name": "stories-components-inputs-helpertext--multiline",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;
